src/cargo/ops/cargo_package/vcs.rs

```diff
@@ -248,20 +254,50 @@ fn dirty_files_outside_pkg_root(
     let pkg_root = pkg.root();
     let workdir = repo.workdir().unwrap();
 
+    let mut dirty_files = HashSet::new();
+
     let meta = pkg.manifest().metadata();
     let metadata_paths: Vec<_> = [&meta.license_file, &meta.readme]
         .into_iter()
         .filter_map(|p| p.as_deref())
         .map(|path| paths::normalize_path(&pkg_root.join(path)))
         .collect();
 
-    let mut dirty_files = HashSet::new();
+    // Unlike other files, lockfile is allowed to be missing,
+    // and can be generated during packaging.
+    // We skip checking when it is missing in both workdir and git index,
+    // otherwise cargo will fail with git2 not found error.
+    let lockfile_path = ws.lock_root().as_path_unlocked().join(LOCKFILE_NAME);
+    let lockfile_path = if lockfile_path.exists() {
+        Some(lockfile_path)
+    } else if let Ok(rel_path) = paths::normalize_path(&lockfile_path).strip_prefix(workdir) {
+        // We don't canonicalize here because non-existing path can't be canonicalized.
+        match repo.status_file(&rel_path) {
+            Ok(s) if s != git2::Status::CURRENT => {
+                dirty_files.insert(lockfile_path);
+            }
+            // Unmodified
+            Ok(_) => {}
+            Err(e) => {
+                debug!(
+                    "check git status failed for `{}` in workdir `{}`: {e}",
+                    rel_path.display(),
+                    workdir.display(),
+                );
+            }
+        }
+        None
+    } else {
+        None
+    };
+
     for rel_path in src_files
         .iter()
         .filter(|p| p.is_symlink_or_under_symlink())
         .map(|p| p.as_ref().as_path())
         .chain(metadata_paths.iter().map(AsRef::as_ref))
         .chain([ws.root_manifest()])
+        .chain(lockfile_path.as_deref().into_iter())
```

Comment:
I'm trying to think through how this PR might be disruptive to people's workflows and the most likely case I can come ip with is people being sloppy with version bumps and publishing which can run into #5979.